### PR TITLE
Add opt-in test-only configuration to Proxy

### PR DESCRIPTION
**Pull request by robobario**
### Type of change

- Enhancement / new feature

### Description

1. Adds a `development` configuration map to the top level of the proxy configuration
2. Adds the concept of features, requiring the user to set `KROXYLICIOUS_UNLOCK_TEST_ONLY_CONFIGURATION=true` by env var or system prop to use the development features. It initially disallows any configuration to be provided. The proxy will not start if the `development` object is non-empty and the proxy does not have this feature enabled (the default)

### Additional Context

Split out of #1635

To enable integration testing the ApiVersions downgrade response with a real client we have an issue. The proxy may be on the latest kafka-clients version, so it's hard to test this case where the client is ahead. To test this we want to add a development only configuration to enable us to control the maximum version per ApiKey supported at the proxy, to masquerade as a proxy that only understands up to a certain version of ApiVersions.

We want to make sure it's clearly not a production feature and force users to jump through some opt-in hoops to enable the functionality.

This PR adds the foundation for this, introducing unlockable features around config loading.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Update documentation
- [ ] Make sure all unit/integration tests pass
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [ ] Ensure the PR references relevant issue(s) so they are closed on merging.
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

### CODE REVIEW

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/ApiVersionsIT.java
Comment (k-wall) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/ApiVersionsIT.java:37]: This is not really part of this PR

**Review by k-wall — APPROVED** [review]

**Comment by k-wall** [comment]
Looks reasonable to me.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java:35]: I can't help but think insisting that all development mode options are expressed at the top level seems overly restrictive. 

What if a filter wants to use it wants it own development configuration? I'm wondering if we actually want to an annotation (`@DevelopmentMode`) that could be applied anywhere on the config model.   The proxy, in production mode, would then walk the configuration tree and check for any @DevelopmentMode annotated fields that were non-null (or Optional not empty).

We could also use the same configuration tree walker to report on use of any deprecated configuration items in a standard way.

WDYAT? If we decide to merge #1648 as it stands (to break the log jam) that's fine too..

Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java:35]: Keen to merge this as-is and get the branches in. 

Definitely makes sense to have it available anywhere in the tree so you don't have to cobble together nested structures in this `development` object. But I think we should add it later.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/ApiVersionsIT.java
Comment (robobario) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/ApiVersionsIT.java]: will move to another PR

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java
Comment (SamBarker) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java:37]: I wonder if we should reduce this to a `warn` and just have the runtime operate in production mode and *ignore* the development configuration. That feels like a more flexible model which would allow people to promote a config file between environments. 

E.g. run DEV/Integration environments in `DEVELOPMENT` mode for some additional validation and then promote that configuration to UAT/Stage/Prod in `PRODUCTION` mode.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java
Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java:35]: I think the way we should do this is:

* Generate the proxy config from schema
* Allow unknown properties on all objects in the schema
* Run a visitor over the configuration instance to check whether there are any unknown properties
* In normal mode we fail if we there are unknown properties, but in normal mode we allow it with a warning logged about the configuration using non-supported properties which have no compatibility guarantees and may disappear without notice in the future.

There are a few choices to be made in exactly how we want that to work. 

I'm not sure I'd call this development mode. That makes it sound kinda _interesting_ to developers, which is the opposite of what we want. It's also ambiguous about what kind of developer (filter author, someone working on kroxy itself, but also our users might also identify with developer). So to me this is more of a `-Dunlock_system_test_only_configs`. 

I guess I'm OK with this merging this PR to unblock, but can we add some wording about "supported properties which have no compatibility guarantees" to the warning we log?

Comment (SamBarker) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java:35]: I share @k-wall's concern about the usefulness of only having development options at the root however its easy to relax restrictions later when we have a use case or requirement for it rather than just guessing it might be useful.

**Review by SamBarker — COMMENTED** [review]
My only hesitation around merging this as is to unblock things is:

Is the mode really independent of the configuration model? Should it just be a synthetic node injected into the config model? 

That might make for a better API surface for plug-ins / filters to consume from and thus avoid the need for knowing about a development config node...

One of the bigger drawbacks to that idea is that it becomes up to the plug-ins to determine what is or isn't legal to load in a given mode and the proxy can't manage or validate that.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java:35]: Happy to make it more specific.

The `unlock_*` pattern would point to a slightly different implementation, feature based. So where I'm passing around a mode it would be more like a `Set<Feature>`.

Comment (SamBarker) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java:35]: > Happy to make it more specific.
> 
> The `unlock_*` pattern would point to a slightly different implementation, feature based. So where I'm passing around a mode it would be more like a `Set<Feature>`.

Does that just suggest we should be using a feature flag library? Such as https://openfeature.dev/

Comment (tombentley) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/ProxyMode.java:35]: Personally I've always struggled to get away from the feeling that open feature is just over engineering. It's certainly too soon to decide something like that when the problem we're trying to solve right now is one where we have single use case for needing to configure something like this in very specific circumstances.

**Comment by robobario** [comment]
> Is the mode really independent of the configuration model? Should it just be a synthetic node injected into the config model?

Configuration around how/what gets loaded from the configuration file feels independent.

We'd also need to add to the Filter API to expose it, as they are currently only aware of their individual configuration objects. I don't think we would want to expose the whole configuration tree to each Filter so it leans towards an independent style anyway.

For now I propose:

1. make the naming more specific since development mode has some baggage in the naming
2. make sure the logging talks about how the configs are unsupported and have no compatibility guarantees
3. leave it harsh/restrictive for now, since the only usage is this testing case I think it's okay to prevent startup

Since we have no compatibility guarantees it should be quick to rewrite/evolve if we want to take this mechanism further.

File: kroxylicious-integration-test-support/src/test/java/io/kroxylicious/test/tester/DefaultKroxyliciousTesterTest.java
Comment (SamBarker) [inline kroxylicious-integration-test-support/src/test/java/io/kroxylicious/test/tester/DefaultKroxyliciousTesterTest.java:540]: Swapping the var order is annoying and means we can't use a method reference...

**Comment by robobario** [comment]
pushed up, the implementation is now in terms of `Set<Feature>` and the feature is unlocked with `-DKROXYLICIOUS_UNLOCK_LOAD_TEST_CONFIGURATION=true`

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java
Comment (SamBarker) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java:101]: This feels a bit odd given you have `Feature.maybeLogWarning`. Why don't we add a flag to the feature like `TestOnly` and make this something like

```
var errors = features.filter(feature.supports(development.get()).findAll().toList();
log.error(errors);
throw new ICE(errors);
```

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/Feature.java
Comment (SamBarker) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/Feature.java:7]: I think this amounts to a public API if its in this package. Maybe it should live in `io.kroxylicious.proxy.internal.config` or something similar?

**Review by SamBarker — APPROVED** [review]
Lets unblock things

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java]: alright I've refactored to something like this, the existing `Set<Feature>` didn't allow for this kind of implementation.

I've changed it so there is now a `Features` class that knows the `enabled/disabled` state for every Feature and the Feature is responsible for applying the validation logic. I think it's tidier than using set contains.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/IllegalConfigurationException.java
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/IllegalConfigurationException.java:7]: I think this would be better in the config subpackage.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java:97]: extract a variable and use it twice.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/KafkaProxy.java:85]:
```suggestion
    public KafkaProxy(@NonNull PluginFactoryRegistry pfr, @NonNull Configuration config, @NonNull Features features) {
```

**Review by k-wall — APPROVED** [review]

**Comment by k-wall** [comment]
Could the coverage sonarcloud be addressed before we merge?

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/config/Feature.java
Comment (SamBarker) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/config/Feature.java:37]: not sure I understand the cardinality here, why does this return a Stream of values rather than just a Optional?

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/config/Features.java
Comment (SamBarker) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/config/Features.java:25]:
```suggestion
    private static final Features ENABLED_BY_DEFAULT = new Features(Map.of());
```
Though maybe thats a bit funky as it implies the proxy does nothing by default...

**Review by SamBarker — APPROVED** [review]
LGTM

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/config/Features.java
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/config/Features.java:25]: Each Feature knows if it is enabled by default or not, so you can have a Feature that is default off and has to be enabled (like test-only configuration), but you could also have a feature that is enabled by default.

The default set of features uses the default enabled-ness of each feature.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/config/Feature.java
Comment (robobario) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/config/Feature.java:37]: I was thinking you could find multiple reasons a configuration isn't supported.

**Pull request merged**
